mail_queue_job/models/mail_mail.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2018 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+from odoo.addons.queue_job.job import job, DEFAULT_PRIORITY
+
+
+class MailMail(models.Model):
+    _inherit = 'mail.mail'
+
+    mail_job_priority = fields.Integer(
+        oldname='priority',
+        default=DEFAULT_PRIORITY,
+        help="Send the email with a priority level.\n"
+             "0 being the higher priority. Default is %d" % DEFAULT_PRIORITY,
+    )
+
+    @api.multi
+    @job(default_channel='root.mail')
+    def _send_mail_jobified(self):
+        """
+        Send the email if
+        - the state is 'outgoing'
+        - the mail exists.
+        :return: str
+        """
+        self.ensure_one()
+        if not self.exists():
+            return "mail.mail record (id=%d) no longer exists" % self.id
+        elif self.state != 'outgoing':
+            return "Not in Outgoing state, ignoring"
+        self.send(auto_commit=False, raise_exception=True)
+        return ""
+
+    @api.multi
+    def _send_email_delay(self, operation='write'):
+        """
+        Execute the _send_mail_jobified function with the priority defined
+        into the mail.mail (self) and a description depending on the
+        operation.
+        This function split given records to create one job queue per mail.
+        The identity_key (on the with_delay function) is very helpful to ensure
+        that another queue job will send the same email.
```

Comment:
I'm not sure what the intention here but beware of this: it won't prevent a concurrent transaction to create a job for the same email at the same time if the concurrent transaction happens within the same timeframe. It will however prevent the creation of a job at a different moment (only before the job is done) or within the same transaction. Is it enough for this use case?

Reply:
The intention is to avoid sending the same `mail.mail` more than once (if a job is already planned for a `mail.mail` recordset, it shouldn't have another job who'll re-send the same `mail.mail`).
But when the job is finished, a user can try to re-sent the same `mail.mail` so it could create a new job for that.
About your first case (same timeframe and concurrent transaction) it should never happens because jobs are created into `create`/`write` so it's the same transaction.
Thank you for your review @guewen